consensus/core/src/network/tonic_network.rs

```diff
@@ -539,26 +537,121 @@ impl<S: NetworkService> NetworkManager<S> for TonicManager {
                     .max_encoding_message_size(config.message_size_limit)
                     .max_decoding_message_size(config.message_size_limit),
             )
-            .serve_with_shutdown(own_address, async move {
-                match rx.await {
-                    Ok(()) => {
-                        debug!("Consensus tonic server is shutting down");
-                    }
-                    Err(e) => {
-                        warn!("Consensus tonic server is shutting down at {own_address}: {e:?}");
-                    }
-                }
-            });
+            .into_service();
+
+        let mut http = Http::new();
+        http.http2_only(true);
+        let http = Arc::new(http);
+
+        let tls_server_config =
+            create_rustls_server_config(&self.context, self.network_keypair.clone());
+        let tls_acceptor = TlsAcceptor::from(Arc::new(tls_server_config));
+
+        // TODO: configure socket options via TcpSocket.
+        let listener = TcpListener::bind(own_address)
+            .await
+            .unwrap_or_else(|e| panic!("Cannot bind to {own_address}: {e:?}"));
+
+        let connections_info = Arc::new(ConnectionsInfo::new(self.context.clone()));
 
         self.server.spawn(async move {
-            if let Err(e) = server.await {
-                warn!("Tonic server failed: {e:?}");
-            } else {
-                info!("Tonic server stopped");
+            let mut connection_handlers = JoinSet::new();
```

Comment:
Probably not for now, but I am thinking whether it would worth somehow guarding this and ensure it won't get saturated by a peer